### PR TITLE
Android Camera GL View

### PR DESCRIPTION
**Pull request by apavlenko**
A new GLES based Camera bridge for getting camera preview as OpenGL texture.
Tutorial-4 refactored using this stuff.

```
cmake.exe -Wno-dev -GNinja -DCMAKE_MAKE_PROGRAM="path\to\ninja\ninja.exe" -DCMAKE_TOOLCHAIN_FILE=../opencv/platforms/android/android.toolchain.cmake -DANDROID_ABI="armeabi-v7a with NEON" -DANDROID_SDK_TARGET=21 -DANDROID_NATIVE_API_LEVEL=14 -DCMAKE_BUILD_WITH_INSTALL_RPATH=ON -DBUILD_ANDROID_EXAMPLES=ON -DINSTALL_ANDROID_EXAMPLES=ON -DWITH_OPENCL=YES -DANDROID_OPENCL_SDK=path\to\OpenCL ../opencv
```
where `path\to\OpenCL` folders contains OpenCL-1.2 headers in `include` and `libOpenCL.so` from the target device in `lib\${ABI_NAME}`

### CODE REVIEW

File: modules/java/generator/src/java/android+CameraGLRendererBase.java [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+            GLES20.glDeleteShader(fshader);
+            fshader = 0;
+            return 0;
+        }
+
+        int program = GLES20.glCreateProgram();
+        GLES20.glAttachShader(program, vshader);
+        GLES20.glAttachShader(program, fshader);
+        GLES20.glLinkProgram(program);
+        Log.d("CameraGLRendererBase", "shaders were compiled OK");
```
Comment (ilya-lavrenov) [inline modules/java/generator/src/java/android+CameraGLRendererBase.java]: `shaders were compiled OK` is true, but you link the shaders and don't check link status...

File: samples/android/tutorial-4-opencl/jni/CLprocessor.cpp [hunk 2]
```diff
@@ -219,3 +230,60 @@ void procOCL_OCV(int texIn, int texOut, int w, int h)
...
+void drawFrameProcCPU(int w, int h, int texOut)
+{
+    LOGD("Processing on CPU");
+    int64_t t;
+
+    // let's modify pixels in FBO texture in C++ code (on CPU)
+    const int BUFF_SIZE = 1<<24;//2k*2k*4;
+    static char tmpBuff[BUFF_SIZE];
+    if(w*h > BUFF_SIZE)
+    {
```
Comment (ilya-lavrenov) [inline samples/android/tutorial-4-opencl/jni/CLprocessor.cpp]: em, why don't you read directly to `cv::Mat`. it allows to avoid ugly `tmpBuff`

**Comment by alalek** [comment]
:+1:

**Comment by PhilLab** [comment]
@apavlenko :+1: Makes Android+OpenCV+OpenGL integration even more easier

**Comment by PhilLab** [comment]
@apavlenko Maybe refactoring the tutorial http://docs.opencv.org/master/d7/dbd/tutorial_android_ocl_intro.html is missing in this pull request

**Pull request merged**